### PR TITLE
changes for Clang and yasm

### PR DESCRIPTION
**Pull request by wjwithagen**
- Clang does not need all flags and options
- uname -m on FreeBSD returns amd64

### CODE REVIEW

**Comment by wjwithagen** [comment]
@tchaikov 
Another small set of commits to get Clang and Yasm going.

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -17,37 +17,58 @@ set(prefix ${CMAKE_INSTALL_PREFIX})
...
+  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wno-gnu-designator -Wno-mismatched-tags")
+  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wno-missing-braces -Wno-parentheses -Wno-deprecated-register")
+endif(CMAKE_CXX_COMPILER_ID MATCHES "GNU")
 
 if(NOT CMAKE_BUILD_TYPE)
   set(CMAKE_BUILD_TYPE "RelWithDebInfo" CACHE STRING "Default BUILD_TYPE is RelWithDebInfo, other options are: Debug, Release, and MinSizeRel." FORCE) 
 endif()
 
-if(NOT CMAKE_BUILD_TYPE STREQUAL Debug)
+if(NOT CMAKE_BUILD_TYPE MATCHES "Debug")
```
Comment (tchaikov) [inline src/CMakeLists.txt]: @wjwithagen would you please move this change out of this commit? you can always change the `CFLAGS`, `CXXFLAGS` and `CMAKE_CXX_FLAGS_DEBUG` out of cmake.

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -17,37 +17,58 @@ set(prefix ${CMAKE_INSTALL_PREFIX})
...
-endif()
-endif()
-CHECK_C_COMPILER_FLAG(-fstack-protector-strong HAS_STACK_PROTECT)
-if (HAS_STACK_PROTECT)
-  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -fstack-protector-strong")
-endif()
+if(CMAKE_CXX_COMPILER_ID MATCHES "GNU")
+  CHECK_C_COMPILER_FLAG("-Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=2" HAS_FORTIFY_SOURCE)
+  if(NOT CMAKE_BUILD_TYPE MATCHES "Debug")
+  if(HAS_FORTIFY_SOURCE)
```
Comment (tchaikov) [inline src/CMakeLists.txt]: could you take this chance to indent this `if` block?

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -144,31 +168,34 @@ try_compile(INTEL_SSE4_1 ${CMAKE_BINARY_DIR} ${sse_srcs}
...
-  if(ARM_NEON)
-    set(ARM_NEON_FLAGS "-mfpu=neon -DARM_NEON")
-  else(ARM_NEON)
-    set(ARM_NEON_FLAGS "-march=armv8-a+simd -DARCH_AARCH64 -DARM_NEON")
-  endif(ARM_NEON)
-else(ARM_NEON OR ARM_NEON2)
-  message(STATUS "Skipping target ec_jerasure_neon & ec_shec_neon: Architecture not ARM")
-endif(ARM_NEON OR ARM_NEON2)
+
+if(NOT arch MATCHES "x84_64|amd64")
```
Comment (tchaikov) [inline src/CMakeLists.txt]: and could you move this change into another commit? and fwiw, this commit will conflict with #10427.

File: src/CMakeLists.txt [hunk 4]
```diff
@@ -70,6 +67,9 @@ if(CMAKE_CXX_COMPILER_ID MATCHES "GNU")
   endif()
 endif(CMAKE_CXX_COMPILER_ID MATCHES "GNU")
 
+set(CMAKE_ASM_COMPILER  ${PROJECT_SOURCE_DIR}/src/yasm-wrapper)
+set(CMAKE_ASM_FLAGS "-f elf64")
+
```
Comment (tchaikov) [inline src/CMakeLists.txt]: could you explain a little bit the rationale of the changes in your commit message?

File: src/CMakeLists.txt [hunk 5]
```diff
@@ -223,8 +223,13 @@ else(INTEL_SSE4_1)
   message(STATUS "Skipping target ec_jerasure_sse4 & ec_shec_sse4: -msse4.1 not supported")
 endif(INTEL_SSE4_1)
 
+set(EXTRALIBS rt ${CMAKE_DL_LIBS} ${ATOMIC_OPS_LIBRARIES})
+if(LINUX)
+  set(LIB_RESOLV resolv)
+  set(LIB_UUID uuid)
```
Comment (tchaikov) [inline src/CMakeLists.txt]: i don't think we are using libuuid anymore. see https://github.com/ceph/ceph/pull/10391/commits/3170f16e864e257416bccdd73777bff775767a8f i am planning to drop it in cmake.

File: src/CMakeLists.txt [hunk 6]
```diff
@@ -623,12 +621,20 @@ if(${WITH_RADOSGW})
...
+endif(WITH_RADOSGW)
+if(WITH_RBD)
   set(DENCODER_EXTRALIBS
       ${DENCODER_EXTRALIBS}
       rbd_replay_types)
-endif(${WITH_RBD})
+  list(APPEND dencoder_srcs
+      $<TARGET_OBJECTS:krbd_objs>)
+endif(WITH_RBD)
+if(WITH_KVS)
```
Comment (tchaikov) [inline src/CMakeLists.txt]: why?

File: src/CMakeLists.txt [hunk 7]
```diff
@@ -223,8 +223,13 @@ else(INTEL_SSE4_1)
   message(STATUS "Skipping target ec_jerasure_sse4 & ec_shec_sse4: -msse4.1 not supported")
 endif(INTEL_SSE4_1)
 
+set(EXTRALIBS rt ${CMAKE_DL_LIBS} ${ATOMIC_OPS_LIBRARIES})
```
Comment (tchaikov) [inline src/CMakeLists.txt]: could add a a short note in the commit message explaining why we want to exclude these libs in non linux build.

File: src/CMakeLists.txt (hunk 1 above)
Comment (wjwithagen) [inline src/CMakeLists.txt]: @tchaikov 
I only added the part where we'd like to run debug, and then these flags look rather appropriate.
But you ar right that they could be set from outside of Cmake.

File: src/CMakeLists.txt (hunk 3 above)
Comment (wjwithagen) [inline src/CMakeLists.txt]: @tchaikov 
Yup, I saw other PR's as well, Will tke it out, and see what will be inserted otherwise

File: src/CMakeLists.txt (hunk 5 above)
Comment (wjwithagen) [inline src/CMakeLists.txt]: @tchaikov 
Feel free to drop it, as FreeBSD does not need it, hence the Linux block around it.
But until that moment, using uuid during linking uses an external lib, and cause breakage
Which I would need to install as a package first.

File: src/CMakeLists.txt (hunk 6 above)
Comment (wjwithagen) [inline src/CMakeLists.txt]: @tchaikov 
For now I chose to put it under WITH_RBD, but krbd is even more tricky since that is never going to run on FreeBSD. It is all Linux code.
So perhaps even wrap it in if(LINUX)??
But I was going to delay that until I started working on RBD in general.

Comment (tchaikov) [inline src/CMakeLists.txt]: sorry, i was asking why we needed to guard `$<TARGET_OBJECTS:parse_secret_objs>` with `WITH_KVS`.

File: src/CMakeLists.txt (hunk 1 above)
Comment (tchaikov) [inline src/CMakeLists.txt]: could you revert this change?

File: src/CMakeLists.txt [hunk 8]
```diff
@@ -637,9 +667,14 @@ install(TARGETS ceph-mon DESTINATION bin)
 if(NOT ALLOCATOR STREQUAL "jemalloc")
   set(disable_jemalloc "DISABLE_JEMALLOC=1")
 endif()
+set(EXTRA_CXXFLAGS -fPIC -Wno-unused-variable -DNDEBUG)
+if(FREEBSD)
+  # set(EXTRA_CXXFLAGS ${EXTRA_CXXFLAGS} -I/usr/local/include)
+  set(MAKEFLAGS ${MAKEFLAGS})
```
Comment (tchaikov) [inline src/CMakeLists.txt]: again, what is this for? it's just a no-op, imo.

File: src/CMakeLists.txt [hunk 9]
```diff
@@ -637,9 +667,14 @@ install(TARGETS ceph-mon DESTINATION bin)
 if(NOT ALLOCATOR STREQUAL "jemalloc")
   set(disable_jemalloc "DISABLE_JEMALLOC=1")
 endif()
+set(EXTRA_CXXFLAGS -fPIC -Wno-unused-variable -DNDEBUG)
+if(FREEBSD)
+  # set(EXTRA_CXXFLAGS ${EXTRA_CXXFLAGS} -I/usr/local/include)
```
Comment (tchaikov) [inline src/CMakeLists.txt]: could you remove this line?

File: src/CMakeLists.txt [hunk 10]
```diff
@@ -637,9 +667,14 @@ install(TARGETS ceph-mon DESTINATION bin)
 if(NOT ALLOCATOR STREQUAL "jemalloc")
   set(disable_jemalloc "DISABLE_JEMALLOC=1")
 endif()
+set(EXTRA_CXXFLAGS -fPIC -Wno-unused-variable -DNDEBUG)
```
Comment (tchaikov) [inline src/CMakeLists.txt]: again, why `-DNDEBUG`?

File: src/CMakeLists.txt (hunk 6 above)
Comment (wjwithagen) [inline src/CMakeLists.txt]: parse_secret_objs has elements that are in keyutils, and that is not available

File: src/CMakeLists.txt (hunk 10 above)
Comment (wjwithagen) [inline src/CMakeLists.txt]: @tchaikov 
Sorry, I'm a bit thick in the head today.

Because there is Linux-ism code used when NDEBUG is off. 
But I'll check with the newly import rocksdb sha.
On the other hand if NDEBUG is defined for the project, is is silly to actually compile rocksdb with NDEBUG.
I'll move a beter version of this to #10418 as well

File: src/CMakeLists.txt (hunk 1 above)
Comment (tchaikov) [inline src/CMakeLists.txt]: @wjwithagen ping?

File: src/CMakeLists.txt (hunk 4 above)
Comment (tchaikov) [inline src/CMakeLists.txt]: @wjwithagen ping?

File: src/CMakeLists.txt (hunk 1 above)
Comment (wjwithagen) [inline src/CMakeLists.txt]: @tchaikov 
It should be out by a long time, but I'll check and otehrwise fix

File: src/CMakeLists.txt [hunk 11]
```diff
@@ -35,19 +51,21 @@ if(NOT CMAKE_BUILD_TYPE STREQUAL Debug)
       CMAKE_C_FLAGS_${build_type_upper})
     string(REGEX REPLACE "(^| )[/-]D *NDEBUG($| )" " " "${flags}" "${${flags}}")
   endforeach()
-endif()
+endif(NOT CMAKE_BUILD_TYPE STREQUAL Debug)
```
Comment (tchaikov) [inline src/CMakeLists.txt]: could you drop unrelated changes?

File: src/CMakeLists.txt [hunk 12]
```diff
@@ -35,19 +51,21 @@ if(NOT CMAKE_BUILD_TYPE STREQUAL Debug)
...
-  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=2")
-endif()
-endif()
-CHECK_C_COMPILER_FLAG(-fstack-protector-strong HAS_STACK_PROTECT)
-if (HAS_STACK_PROTECT)
-  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -fstack-protector-strong")
-endif()
+if(CMAKE_CXX_COMPILER_ID MATCHES "GNU")
+  CHECK_C_COMPILER_FLAG("-Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=2" HAS_FORTIFY_SOURCE)
+  if(NOT CMAKE_BUILD_TYPE MATCHES "Debug")
```
Comment (tchaikov) [inline src/CMakeLists.txt]: again, i see you'd prefer "MATCHES", but can we be consistent, and use `STREQUAL`, if we don't need to match the string in question with a regexp?

File: src/CMakeLists.txt [hunk 13]
```diff
@@ -476,7 +496,16 @@ set(libcommon_files
   ${arch_files}
   ${auth_files}
   $<TARGET_OBJECTS:compressor_objs>
-  ${mds_files})
+  ${mds_files}
```
Comment (tchaikov) [inline src/CMakeLists.txt]: drop unrelated changes, if you please.

File: src/CMakeLists.txt [hunk 14]
```diff
@@ -476,7 +496,16 @@ set(libcommon_files
...
+  )
+if(LINUX)
```
Comment (tchaikov) [inline src/CMakeLists.txt]: and this change is not related to clang or yasm.

File: src/CMakeLists.txt [hunk 15]
```diff
@@ -476,7 +496,16 @@ set(libcommon_files
...
+  list(APPEND libcommon_files msg/async/EventEpoll.cc)
+  message(STATUS " Using EventEpoll for events.")
+elseif(FREEBSD)
```
Comment (tchaikov) [inline src/CMakeLists.txt]: OSX (macOS) also uses kqueue.

File: src/CMakeLists.txt [hunk 16]
```diff
@@ -17,11 +17,27 @@ set(prefix ${CMAKE_INSTALL_PREFIX})
 
 add_definitions("-DCEPH_LIBDIR=\"${CMAKE_INSTALL_FULL_LIBDIR}\"")
 add_definitions("-DCEPH_PKGLIBDIR=\"${CMAKE_INSTALL_FULL_PKGLIBDIR}\"")
-add_definitions("-DHAVE_CONFIG_H -D__CEPH__ -D_FILE_OFFSET_BITS=64 -D_REENTRANT -D_THREAD_SAFE -D__STDC_FORMAT_MACROS -D_GNU_SOURCE")
+add_definitions("-DHAVE_CONFIG_H -D__CEPH__ -D_REENTRANT -D_THREAD_SAFE -D__STDC_FORMAT_MACROS")
```
Comment (tchaikov) [inline src/CMakeLists.txt]: could you cleanup the commit message to remove the breadcrumbs from it?

File: src/CMakeLists.txt [hunk 17]
```diff
@@ -552,6 +555,21 @@ endif(HAVE_ARMV8_CRC)
 
 add_library(common_utf8 STATIC common/utf8.c)
 
+<<<<<<< 64564d7b5367df64622d693180a21d07c33e4262
```
Comment (tchaikov) [inline src/CMakeLists.txt]: please resolve this merge conflict.

File: src/CMakeLists.txt [hunk 18]
```diff
@@ -637,12 +620,22 @@ if(${WITH_RADOSGW})
...
   )
-endif(${WITH_RADOSGW})
-if(${WITH_RBD})
+endif(WITH_RADOSGW)
+if(WITH_RBD)
   set(DENCODER_EXTRALIBS
       ${DENCODER_EXTRALIBS}
       rbd_replay_types)
-endif(${WITH_RBD})
+  if(LINUX)
```
Comment (tchaikov) [inline src/CMakeLists.txt]: i am confused. do we have a combination of RBD + non-LINUX?

File: src/CMakeLists.txt [hunk 19]
```diff
@@ -637,12 +620,22 @@ if(${WITH_RADOSGW})
...
+    list(APPEND dencoder_srcs
+        $<TARGET_OBJECTS:krbd_objs>)
+  endif(LINUX)
+endif(WITH_RBD)
+if(WITH_KVS)
```
Comment (tchaikov) [inline src/CMakeLists.txt]: i still don't understand the relation between `WITH_KVS` and `parse_secret_objs`. could you enlighten me on it?

**Comment by wjwithagen** [comment]
@tchaikov 
I'm going to rip out the whole commit with the RBD and KVS stuff
And fix the CLANG/yasm part, so we can at least get that one in.
For the RBD and KVS I'll either add it to one of the other pulls, or even make it a seperate one.

File: src/CMakeLists.txt (hunk 11 above)
Comment (wjwithagen) [inline src/CMakeLists.txt]: @tchaikov 
I was under the impression that "while there" one is allowed to fix imperfacetions in
code?
But I'll take it out.

File: src/CMakeLists.txt (hunk 12 above)
Comment (wjwithagen) [inline src/CMakeLists.txt]: @tchaikov 
Yup, I'd definitly like the more liberal approach to coding, because I also consider it "defensive" programming in this case. If things we're to be ported to yet another compiler, is is just adding a expression to the reg_expr. So we are already prepared.
But I'll put it back.

**Comment by tchaikov** [comment]
could you address the comments? they are dropped on the floor, i guess.

**Comment by wjwithagen** [comment]
@tchaikov 
It is not my intention to drop anything on the floor.
I'm having sort of a fight with git rebase to get ride of a merge confilict.

File: src/CMakeLists.txt (hunk 18 above)
Comment (wjwithagen) [inline src/CMakeLists.txt]: @@tchaikov
Not yet, but it is certainly my intention to do so.

File: src/CMakeLists.txt (hunk 19 above)
Comment (wjwithagen) [inline src/CMakeLists.txt]: @tchaikov 
Could be that I'm misinterpreting things. But the KeyValue store seems to use the file common/secret.c which includes keyutils.h.
And that we don't have in FreeBSD.
And I have not figured out how to patch this up: either write a replacement, and/or stub it up.

I'll take a note of this and put it at least in my freebsd-todo.txt

**Comment by wjwithagen** [comment]
@tchaikov 
Failure is in eph_objectstore_tool.py, which seems to be in a real bad state.

File: src/CMakeLists.txt (hunk 16 above)
Comment (tchaikov) [inline src/CMakeLists.txt]: @wjwithagen could you remove 
```
 - The ARM dependant code will need some care as well, but that is
   in the works in other PR's
 - Rocksdb rework is also moved to another PR
```

Comment (tchaikov) [inline src/CMakeLists.txt]: @wjwithagen could you remove 
```
 - The ARM dependant code will need some care as well, but that is
   in the works in other PR's
 - Rocksdb rework is also moved to another PR
```

in the commit message?

File: src/CMakeLists.txt [hunk 20]
```diff
@@ -94,8 +112,11 @@ else(no_yasm)
   endif(arch STREQUAL "x86_64")
 endif(no_yasm)
 
-
-set(CMAKE_CXX_FLAGS "${CMAKE_C_FLAGS} -ftemplate-depth-1024 -Wnon-virtual-dtor -Wno-invalid-offsetof -Wstrict-null-sentinel -Woverloaded-virtual")
+set(CMAKE_CXX_FLAGS "${CMAKE_C_FLAGS} -ftemplate-depth-1024 -Wno-invalid-offsetof")
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wnon-virtual-dtor -Woverloaded-virtual")
+if(CMAKE_CXX_COMPILER_ID MATCHES "GNU")
```
Comment (tchaikov) [inline src/CMakeLists.txt]: would be better if we can be consistent regarding to the use of `MATCHES "GNU"` and `STREQUAL GNU`?

File: src/CMakeLists.txt (hunk 19 above)
Comment (tchaikov) [inline src/CMakeLists.txt]: > But the KeyValue store seems to use the file common/secret.c which includes keyutils.h.

`secret.c` is used by `mount.ceph.c` and `krbd.cc` also. i am not sure how you come to this.

File: src/CMakeLists.txt (hunk 15 above)
Comment (tchaikov) [inline src/CMakeLists.txt]: @wjwithagen ping?

File: src/CMakeLists.txt (hunk 11 above)
Comment (tchaikov) [inline src/CMakeLists.txt]: yes, 1) but i don't think we need to "fix" it. the `if` block is smaller enough, that we know where it starts. 2) and you are not changing the `if` block, so i don't really think you were there.

File: src/CMakeLists.txt [hunk 21]
```diff
@@ -754,9 +782,11 @@ install(PROGRAMS
 
 add_subdirectory(bash_completion)
 
+if(WITH_KVS)
```
Comment (tchaikov) [inline src/CMakeLists.txt]: > I'm going to rip out the whole commit with the RBD and KVS stuff
> And fix the CLANG/yasm part, so we can at least get that one in.
> For the RBD and KVS I'll either add it to one of the other pulls, or even make it a seperate one.

@wjwithagen ping?

**Comment by tchaikov** [comment]
please ping me if all comment addressed, thanks.

**Comment by wjwithagen** [comment]
@tchaikov 
Ping

File: src/CMakeLists.txt (hunk 19 above)
Comment (wjwithagen) [inline src/CMakeLists.txt]: On 2-8-2016 09:58, Kefu Chai wrote:
> In src/CMakeLists.txt
> <https://github.com/ceph/ceph/pull/10417#discussion_r73109596>:
> 
>>    set(DENCODER_EXTRALIBS
>>        ${DENCODER_EXTRALIBS}
>>        rbd_replay_types)
>> -endif(${WITH_RBD})
>> +  if(LINUX)
>> +    list(APPEND dencoder_srcs
>> +        $<TARGET_OBJECTS:krbd_objs>)
>> +  endif(LINUX)
>> +endif(WITH_RBD)
>> +if(WITH_KVS)
> 
>     But the KeyValue store seems to use the file common/secret.c which
>     includes keyutils.h.
> 
> |secret.c| is used by |mount.ceph.c| and |krbd.cc| also. i am not sure
> how you come to this.


I just looked at the source of secret.c. :)
Perhaps it can be removed, but have not yet dared to do so.
KV exclusion for this also seems to work.

**Comment by tchaikov** [comment]
lgtm.

**Comment by wjwithagen** [comment]
@tchaikov 
This has been in this state for about 7 days.
Any particulars we are waiting for?

**Comment by tchaikov** [comment]
@wjwithagen lemme push this change to gb to get it tested. then i guess it's good to merge.

**Comment by tchaikov** [comment]
pushed to wip-kefu-testing.

**Comment by wjwithagen** [comment]
@tchaikov 
Thanx for getting this in. Up to the next ones.....
